File: typegraph/python/typegraph/t.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import json as JsonLib
from typing import Any, Dict, List, Optional, Tuple, Union
import copy

from typing_extensions import Self

from typegraph.effects import EffectType
from typegraph.gen.exports.core import (
    FuncParams,
    ParameterTransform,
    TypeList,
    TypeBase,
    TypeEither,
    TypeFile,
    TypeFloat,
    TypeFunc,
    TypeInteger,
    TypeOptional,
    TypeString,
    TypeStruct,
    TypeUnion,
    PolicySpec as WitPolicySpec,
)
from typegraph.gen.exports.runtimes import EffectRead
from typegraph.gen.exports.utils import Reduce
from typegraph.gen.types import Err
from typegraph.graph.typegraph import (
    core,
    store,
    ApplyFromArg,
    ApplyFromContext,
    ApplyFromParent,
    ApplyFromSecret,
    ApplyFromStatic,
)
from typegraph.injection import (
    serialize_generic_injection,
    serialize_parent_injection,
    serialize_static_injection,
)
from typegraph.policy import Policy, PolicyPerEffect, PolicySpec, get_policy_chain
from typegraph.runtimes.deno import Materializer
from typegraph.utils import (
    ConfigSpec,
    build_reduce_data,
    serialize_config,
)
from typegraph.wit import wit_utils

# TODO: better approach?
og_list = list


class typedef:
    id: int
    runtime_config: Optional[List[Tuple[str, str]]]
    injection: Optional[str]
    policy_chain: Optional[List[WitPolicySpec]]
    name: Optional[str]

    def __init__(self, id: int):
        self.id = id
        self.runtime_config = None
        self.injection = None
        self.policy_chain = None
        self.name = None

    def __repr__(self):
        res = core.get_type_repr(store, self.id)
        if isinstance(res, Err):
            raise Exception(res.value)
        return res.value

    def with_policy(self, *policies: Optional[PolicySpec]) -> Self:
        policy_chain = get_policy_chain(policies)
        res = core.with_policy(store, self.id, policy_chain)
        if isinstance(res, Err):
            raise Exception(res.value)

        ret = copy.copy(self)
        ret.id = res.value
        ret.policy_chain = policy_chain
        return ret

    def rename(self, name: str) -> Self:
        res = core.rename_type(store, self.id, name)

        if isinstance(res, Err):
            raise Exception(res.value)

        ret = copy.copy(self)
        ret.id = res.value
        ret.name = name
        return ret

    def _with_injection(self, injection: str) -> Self:
        res = core.with_injection(store, self.id, injection)
        if isinstance(res, Err):
            raise Exception(res.value)

        ret = copy.copy(self)
        ret.id = res.value
        ret.injection = injection
        return ret

    def optional(
        self,
        default_value: Optional[Any] = None,
        config: Optional[ConfigSpec] = None,
    ) -> "optional":
        if isinstance(self, optional):
            return self
        return optional(self, default_item=default_value, config=config)

    def set(self, value: Union[any, Dict[EffectType, any]]):
        return self._with_injection(serialize_static_injection(value))

    def inject(self, value: Union[any, Dict[EffectType, any]]):
        return self._with_injection(serialize_generic_injection("dynamic", value))

    def from_context(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_generic_injection("context", value))

    def from_secret(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_generic_injection("secret", value))

    def from_parent(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_parent_injection(value))

    def from_random(self):
        return self._with_injection(serialize_generic_injection("random", None))


class _TypeWithPolicy(typedef):
    base: "typedef"
    policy: Tuple[Union[Policy, PolicyPerEffect], ...]

    def __init__(
        self,
        id: int,
        base: "typedef",
        policy: Tuple[Union[Policy, PolicyPerEffect], ...],
    ):
        super().__init__(id)
        self.base = base
        self.policy = policy

    def __getattr__(self, name):
        if name == "policy":
            return self.policy
        return getattr(self.base, name)


# self.id refer to the wrapper id
# self.* refer to the base id
class _TypeWrapper(typedef):
    base: "typedef"

    def __init__(
        self,
        id: int,
        base: "typedef",
    ):
        super().__init__(id)
        self.base = base

    def __getattr__(self, name):
        if name == "id":
            return self.id
        return getattr(self.base, name)


class integer(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    exclusive_minimum: Optional[int] = None
    exclusive_maximum: Optional[int] = None
    multiple_of: Optional[int] = None
    enumeration: Optional[List[int]] = None
    as_id: bool

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        exclusive_minimum: Optional[int] = None,
        exclusive_maximum: Optional[int] = None,
        multiple_of: Optional[int] = None,
        enum: Optional[List[int]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        as_id: bool = False,
    ):
        data = TypeInteger(
            min=min,
            max=max,
            exclusive_minimum=exclusive_minimum,
            exclusive_maximum=exclusive_maximum,
            multiple_of=multiple_of,
            enumeration=enum,
        )
        runtime_config = serialize_config(config)
        # raise Exception(runtime_config)
        res = core.integerb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=as_id),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.min = min
        self.max = max
        self.exclusive_minimum = exclusive_minimum
        self.exclusive_maximum = exclusive_maximum
        self.multiple_of = multiple_of
        self.enumeration = enum
        self.runtime_config = runtime_config
        self.as_id = as_id


class float(typedef):
    min: Optional[float] = None
    max: Optional[float] = None
    exclusive_minimum: Optional[float] = None
    exclusive_maximum: Optional[float] = None
    multiple_of: Optional[float] = None
    enumeration: Optional[List[float]] = None

    def __init__(
        self,
        *,
        min: Optional[float] = None,
        max: Optional[float] = None,
        exclusive_minimum: Optional[float] = None,
        exclusive_maximum: Optional[float] = None,
        multiple_of: Optional[float] = None,
        enum: Optional[List[float]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeFloat(
            min=min,
            max=max,
            exclusive_minimum=exclusive_minimum,
            exclusive_maximum=exclusive_maximum,
            multiple_of=multiple_of,
            enumeration=enum,
        )
        runtime_config = serialize_config(config)
        res = core.floatb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.min = min
        self.max = max
        self.exclusive_minimum = exclusive_minimum
        self.exclusive_maximum = exclusive_maximum
        self.multiple_of = multiple_of
        self.enumeration = enum
        self.runtime_config = runtime_config


class boolean(typedef):
    def __init__(
        self, *, name: Optional[str] = None, config: Optional[ConfigSpec] = None
    ):
        runtime_config = serialize_config(config)
        res = core.booleanb(
            store, TypeBase(name=name, runtime_config=runtime_config, as_id=False)
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.runtime_config = runtime_config


class string(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    pattern: Optional[str] = None
    format: Optional[str] = None
    enumeration: Optional[List[str]] = None
    as_id: bool

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        pattern: Optional[str] = None,
        format: Optional[str] = None,
        enum: Optional[List[str]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        as_id: bool = False,
    ):
        enum_variants = None
        if enum is not None:
            enum_variants = og_list(JsonLib.dumps(variant) for variant in enum)

        data = TypeString(
            min=min, max=max, pattern=pattern, format=format, enumeration=enum_variants
        )

        runtime_config = serialize_config(config)
        res = core.stringb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=as_id),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.min = min
        self.max = max
        self.pattern = pattern
        self.format = format
        self.enumeration = enum
        self.runtime_config = runtime_config
        self.as_id = as_id


def uuid(
    *,
    config: Optional[ConfigSpec] = None,
    as_id: bool = False,
    name: Optional[str] = None,
) -> string:
    return string(format="uuid", config=config, as_id=as_id, name=name)


def email(
    *,
    config: Optional[Optional[ConfigSpec]] = None,
    as_id: bool = False,
    name: Optional[str] = None,
) -> string:
    return string(format="email", config=config, as_id=as_id, name=name)


def uri() -> string:
    return string(format="uri")


def ean() -> string:
    return string(format="ean")


def path() -> string:
    return string(format="path")


def date() -> string:
    return string(format="date")


def datetime() -> string:
    return string(format="date-time")


def json() -> string:
    return string(format="json")


def enum(
    variants: List[str],
    name: Optional[str] = None,
):
    return string(enum=variants, name=name)


class file(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    allow: Optional[List[str]]

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        allow: Optional[List[str]] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeFile(
            min=min,
            max=max,
            allow=allow,
        )

        runtime_config = serialize_config(config)
        res = core.fileb(
            store,
            data,
            TypeBase(name=None, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)

        super().__init__(res.value)
        self.min = min
        self.max = max
        self.allow = allow


class list(typedef):
    items: typedef = None
    min: Optional[int] = None
    max: Optional[int] = None
    unique_items: Optional[bool] = None

    def __init__(
        self,
        items: typedef,
        min: Optional[int] = None,
        max: Optional[int] = None,
        unique_items: Optional[bool] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeList(
            of=items.id,
            min=min,
            max=max,
            unique_items=unique_items,
        )

        runtime_config = serialize_config(config)
        res = core.listb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.min = min
        self.max = max
        self.items = items
        self.unique_items = unique_items
        self.runtime_config = runtime_config


class optional(typedef):
    item: Optional[typedef] = None
    default_item: Optional[str] = None

    def __init__(
        self,
        item: typedef,
        default_item: Optional[Any] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeOptional(
            of=item.id,
            default_item=None if default_item is None else JsonLib.dumps(default_item),
        )

        runtime_config = serialize_config(config)
        res = core.optionalb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.item = item
        self.default_item = default_item
        self.runtime_config = runtime_config


class union(typedef):
    variants: List[typedef] = []

    def __init__(
        self,
        variants: List[typedef],
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeUnion(variants=og_list(map(lambda v: v.id, variants)))

        runtime_config = serialize_config(config)
        res = core.unionb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.variants = variants
        self.runtime_config = runtime_config


class either(typedef):
    variants: List[typedef] = []

    def __init__(
        self,
        variants: List[typedef],
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeEither(variants=og_list(map(lambda v: v.id, variants)))

        runtime_config = serialize_config(config)
        res = core.eitherb(
            store,
            data,
            TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.variants = variants
        self.runtime_config = runtime_config


class struct(typedef):
    props: Dict[str, typedef]
    additional_props: bool
    min: Optional[int]
    max: Optional[int]
    enumeration: Optional[List[int]] = None

    def __init__(
        self,
        props: Optional[Dict[str, typedef]] = None,
        *,
        additional_props: bool = False,
        min: Optional[int] = None,
        max: Optional[int] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        enum: Optional[List[Dict[str, Any]]] = None,
    ):
        if self.__class__ != struct:  # custom class
            if len(self.__class__.__bases__) > 1:
                raise Exception("multiple inheritance is currently not supported")
            (base,) = self.__class__.__bases__
            child_cls = self.__class__
            child_attr = set([i for i in vars(child_cls) if not i.startswith("__")])
            parent_attr = set([i for i in vars(base) if not i.startswith("__")])

            # reserved field check
            reserved_attr = set(vars(struct)).union(vars(typedef))
            common = sorted(reserved_attr.intersection(child_attr))
            if len(common) > 0:
                err_msg = ", ".join(common)
                if len(common) == 1:
                    err_msg += " is a reserved field"
                else:
                    err_msg += " are reserved fields"
                raise Exception(err_msg)
            self_attr = child_attr
            if base != struct:
                # child.props should inherit parent.props
                curr_base = base
                while curr_base != struct:
                    if len(curr_base.__bases__) > 1:
                        raise Exception(
                            "multiple inheritance is currently not supported"
                        )
                    (curr_base,) = curr_base.__bases__
                    fields = set([i for i in vars(curr_base) if not i.startswith("__")])
                    parent_attr = parent_attr.union(fields)
                self_attr = self_attr.union(parent_attr)
            props = {}
            for attr in sorted(self_attr):
                value = getattr(self, attr)
                if isinstance(value, typedef):
                    props[attr] = value

        else:
            props = props or {}

        data = TypeStruct(
            props=og_list((name, tpe.id) for (name, tpe) in props.items()),
            additional_props=additional_props,
            min=min,
            max=max,
            enumeration=[JsonLib.dumps(v) for v in enum] if enum else None,
        )

        runtime_config = serialize_config(config)
        res = core.structb(
            store,
            data,
            base=TypeBase(name=name, runtime_config=runtime_config, as_id=False),
        )
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.props = props
        self.runtime_config = runtime_config
        self.enumeration = enum

    def extend(self, props: Dict[str, typedef]):
        return struct(props={**self.props, **props})


ApplyParamObjectNode = Dict[str, "ApplyParamNode"]
ApplyParamArrayNode = List["ApplyParamNode"]
ApplyParamLeafNode = Union[
    ApplyFromArg, ApplyFromStatic, ApplyFromContext, ApplyFromSecret, ApplyFromParent
]

ApplyParamNode = Union[ApplyParamObjectNode, ApplyParamArrayNode, ApplyParamLeafNode]


def serialize_apply_param_node(node: ApplyParamNode) -> Any:
    if isinstance(node, ApplyFromArg):
        return {"source": "arg", "name": node.name}
    if isinstance(node, ApplyFromStatic):
        return {"source": "static", "value": JsonLib.dumps(node.value)}
    if isinstance(node, ApplyFromContext):
        return {"source": "context", "key": node.key}
    if isinstance(node, ApplyFromSecret):
        return {"source": "secret", "key": node.key}
    if isinstance(node, ApplyFromParent):
        return {"source": "parent", "type_name": node.type_name}
    if isinstance(node, dict):
        return {
            "type": "object",
            "fields": {k: serialize_apply_param_node(v) for k, v in node.items()},
        }
    if isinstance(node, (og_list, tuple)):
        return {"type": "array", "items": [serialize_apply_param_node(v) for v in node]}

    raise Exception(f"unexpected node type: node={node}")


class func(typedef):
    inp: Union[struct, typedef]
    out: typedef
    mat: Materializer
    rate_calls: bool
    rate_weight: Optional[int]
    parameter_transform: Optional[ParameterTransform]

    def __init__(
        self,
        inp: Union[struct, typedef],
        out: typedef,
        mat: Materializer,
        /,
        rate_calls: bool = False,
        rate_weight: Optional[int] = None,
        parameter_transform: Optional[ParameterTransform] = None,
    ):
        data = TypeFunc(
            inp=inp.id,
            out=out.id,
            parameter_transform=parameter_transform,
            mat=mat.id,
            rate_calls=rate_calls,
            rate_weight=rate_weight,
        )
        res = core.funcb(store, data)
        if isinstance(res, Err):
            raise Exception(res.value)
        id = res.value
        super().__init__(id)
        self.inp = inp
        self.out = out
        self.mat = mat
        self.rate_calls = rate_calls
        self.rate_weight = rate_weight
        self.parameter_transform = parameter_transform

    def rate(self, calls: bool = False, weight: Optional[int] = None) -> "func":
        return func(
            self.inp,
            self.out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=calls,
            rate_weight=weight,
        )

    def extend(self, props: Dict[str, typedef]):
        res = core.extend_struct(
            store, self.out.id, og_list((k, v.id) for k, v in props.items())
        )
        if isinstance(res, Err):
            raise Exception(res.value)

        out = typedef(res.value)

        return func(
            self.inp,
            out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
        )

    def reduce(self, value: Dict[str, Any]) -> "func":
        data = Reduce(paths=build_reduce_data(value, [], []))
        reduced_id = wit_utils.gen_reduceb(store, self.inp.id, data=data)

        if isinstance(reduced_id, Err):
            raise Exception(reduced_id.value)

        # TODO typedef(...).as_struct()
        return func(
            typedef(id=reduced_id.value),
            self.out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
        )

    def apply(self, value: ApplyParamObjectNode) -> "func":
        serialized = serialize_apply_param_node(value)
        assert isinstance(serialized, dict)
        assert serialized["type"] == "object"
        transform_tree = JsonLib.dumps(serialized["fields"])

        transform_data = core.get_transform_data(store, self.inp.id, transform_tree)
        if isinstance(transform_data, Err):
            import sys

            print(transform_tree, file=sys.stderr)
            raise Exception(transform_data.value)

        return func(
            typedef(transform_data.value.query_input),
            self.out,
            self.mat,
            parameter_transform=transform_data.value.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
        )

    @staticmethod
    def from_type_func(
        data: FuncParams, rate_calls: bool = False, rate_weight: Optional[int] = None
    ) -> "func":
        # Note: effect is a just placeholder
        # in the deno frontend, we do not have to fill the effect attribute on materializers
        mat = Materializer(id=data.mat, effect=EffectRead())
        return func(
            typedef(id=data.inp),
            typedef(id=data.out),
            mat,
            rate_calls=rate_calls,
            rate_weight=rate_weight,
            parameter_transform=None,
        )


def gen(
    out: typedef,
    mat: Materializer,
    rate_calls: bool = False,
    rate_weight: Optional[int] = None,
):
    return func(
        struct({}),
        out,
        mat,
        rate_calls=rate_calls,
        rate_weight=rate_weight,
    )

File: src/typegraph/rust/src/t.rs
use crate::{
    error::Result,
    wasm::{
        self,
        core::{
            FuncParams, ParameterTransform, TypeBase, TypeEither, TypeFloat, TypeFunc, TypeId,
            TypeInteger, TypeList, TypeOptional, TypeString, TypeStruct, TypeUnion,
        },
    },
};

pub trait TypeBuilder: Sized {
    fn build(self) -> Result<TypeId>;

    fn optional(self) -> Result<OptionalBuilder> {
        optional(self)
    }
}

pub trait BaseBuilder: Sized {
    fn base_mut(&mut self) -> &mut TypeBase;

    fn named(mut self, name: String) -> Self {
        self.base_mut().name = Some(name);
        self
    }

    fn as_id(mut self) -> Self {
        self.base_mut().as_id = true;
        self
    }
}

impl TypeBuilder for TypeId {
    fn build(self) -> Result<TypeId> {
        Ok(self)
    }
}

#[allow(clippy::derivable_impls)]
impl Default for TypeBase {
    fn default() -> Self {
        Self {
            name: None,
            runtime_config: None,
            as_id: false,
        }
    }
}

#[allow(clippy::derivable_impls)]
impl Default for TypeInteger {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct IntegerBuilder {
    data: TypeInteger,
    base: TypeBase,
}

impl IntegerBuilder {
    pub fn min(mut self, min: i32) -> Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(mut self, max: i32) -> Self {
        self.data.max = Some(max);
        self
    }

    pub fn x_min(mut self, min: i32) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    pub fn x_max(mut self, max: i32) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }

    pub fn multiple(mut self, n: i32) -> Self {
        self.data.multiple_of = Some(n);
        self
    }

    pub fn enumerate(mut self, values: impl IntoIterator<Item = i32>) -> Self {
        self.data.enumeration = Some(values.into_iter().collect());
        self
    }
}

impl TypeBuilder for IntegerBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_integerb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for IntegerBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn integer() -> IntegerBuilder {
    IntegerBuilder::default()
}

#[allow(clippy::derivable_impls)]
impl Default for TypeFloat {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct FloatBuilder {
    base: TypeBase,
    data: TypeFloat,
}

impl FloatBuilder {
    pub fn min(mut self, min: f64) -> Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(mut self, max: f64) -> Self {
        self.data.max = Some(max);
        self
    }

    pub fn x_min(mut self, min: f64) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    pub fn x_max(mut self, max: f64) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }

    pub fn multiple(mut self, n: f64) -> Self {
        self.data.multiple_of = Some(n);
        self
    }

    pub fn enumerate(mut self, values: impl IntoIterator<Item = f64>) -> Self {
        self.data.enumeration = Some(values.into_iter().collect());
        self
    }
}

impl TypeBuilder for FloatBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_floatb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for FloatBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn float() -> FloatBuilder {
    FloatBuilder::default()
}

#[allow(clippy::derivable_impls)]
impl Default for TypeString {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            format: None,
            pattern: None,
            enumeration: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct StringBuilder {
    base: TypeBase,
    data: TypeString,
}

impl StringBuilder {
    pub fn min(mut self, min: u32) -> Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(mut self, max: u32) -> Self {
        self.data.max = Some(max);
        self
    }

    pub fn format(mut self, format: impl Into<String>) -> Self {
        self.data.format = Some(format.into());
        self
    }

    pub fn enumerate(mut self, values: impl IntoIterator<Item = impl ToString>) -> Self {
        self.data.enumeration = Some(values.into_iter().map(|v| v.to_string()).collect());
        self
    }
}

impl TypeBuilder for StringBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_stringb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for StringBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn string() -> StringBuilder {
    StringBuilder::default()
}

impl Default for TypeOptional {
    fn default() -> Self {
        Self {
            of: u32::MAX,
            default_item: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct OptionalBuilder {
    base: TypeBase,
    data: TypeOptional,
}

impl TypeBuilder for OptionalBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_optionalb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for OptionalBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn optional(id: impl TypeBuilder) -> Result<OptionalBuilder> {
    Ok(OptionalBuilder {
        base: TypeBase::default(),
        data: TypeOptional {
            of: id.build()?,
            default_item: None,
        },
    })
}

impl Default for TypeList {
    fn default() -> Self {
        Self {
            of: u32::MAX,
            min: None,
            max: None,
            unique_items: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct ListBuilder {
    base: TypeBase,
    data: TypeList,
}

impl ListBuilder {
    pub fn min(mut self, min: u32) -> Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(mut self, max: u32) -> Self {
        self.data.max = Some(max);
        self
    }

    pub fn unique(mut self, value: bool) -> Self {
        self.data.unique_items = Some(value);
        self
    }
}

impl TypeBuilder for ListBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_listb(s, self.data, &self.base))
    }
}

impl BaseBuilder for ListBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn list(id: impl TypeBuilder) -> Result<ListBuilder> {
    Ok(ListBuilder {
        base: TypeBase::default(),
        data: TypeList {
            of: id.build()?,
            ..Default::default()
        },
    })
}

#[allow(clippy::derivable_impls)]
impl Default for TypeUnion {
    fn default() -> Self {
        Self {
            variants: Vec::default(),
        }
    }
}

#[derive(Debug, Default)]
pub struct UnionBuilder {
    base: TypeBase,
    data: TypeUnion,
}

impl UnionBuilder {
    pub fn add(mut self, id: impl TypeBuilder) -> Result<Self> {
        self.data.variants.push(id.build()?);
        Ok(self)
    }
}

impl TypeBuilder for UnionBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_unionb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for UnionBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn union(variants: impl IntoIterator<Item = impl TypeBuilder>) -> Result<UnionBuilder> {
    Ok(UnionBuilder {
        data: TypeUnion {
            variants: variants
                .into_iter()
                .map(|id| id.build())
                .collect::<Result<_>>()?,
        },
        ..Default::default()
    })
}

#[allow(clippy::derivable_impls)]
impl Default for TypeEither {
    fn default() -> Self {
        Self {
            variants: Vec::default(),
        }
    }
}

#[derive(Debug, Default)]
pub struct EitherBuilder {
    base: TypeBase,
    data: TypeEither,
}

impl TypeBuilder for EitherBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_eitherb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for EitherBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn either(variants: impl IntoIterator<Item = impl TypeBuilder>) -> Result<EitherBuilder> {
    Ok(EitherBuilder {
        data: TypeEither {
            variants: variants
                .into_iter()
                .map(|id| id.build())
                .collect::<Result<_>>()?,
        },
        ..Default::default()
    })
}

#[allow(clippy::derivable_impls)]
impl Default for TypeStruct {
    fn default() -> Self {
        Self {
            props: Vec::new(),
            additional_props: false,
            min: None,
            max: None,
            enumeration: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct StructBuilder {
    base: TypeBase,
    data: TypeStruct,
}

impl StructBuilder {
    pub fn prop(mut self, name: impl ToString, id: impl TypeBuilder) -> Result<Self> {
        self.data.props.push((name.to_string(), id.build()?));
        Ok(self)
    }

    pub fn props<S, T>(mut self, values: impl IntoIterator<Item = (S, T)>) -> Result<Self>
    where
        S: ToString,
        T: TypeBuilder,
    {
        self.data.props.extend(
            values
                .into_iter()
                .map(|(name, id)| id.build().map(|id| (name.to_string(), id)))
                .collect::<Result<Vec<_>>>()?,
        );
        Ok(self)
    }

    pub fn min(mut self, min: u32) -> Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(mut self, max: u32) -> Self {
        self.data.max = Some(max);
        self
    }
}

impl TypeBuilder for StructBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_structb(s, &self.data, &self.base))
    }
}

impl BaseBuilder for StructBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

pub fn r#struct() -> StructBuilder {
    StructBuilder::default()
}

impl Default for TypeFunc {
    fn default() -> Self {
        Self {
            inp: u32::MAX,
            out: u32::MAX,
            parameter_transform: None,
            mat: u32::MAX,
            rate_calls: false,
            rate_weight: None,
        }
    }
}

#[derive(Debug, Default)]
pub struct FuncBuilder {
    data: TypeFunc,
}

impl FuncBuilder {
    pub fn rate_call(mut self, value: bool) -> Self {
        self.data.rate_calls = value;
        self
    }

    pub fn rate_weight(mut self, value: u32) -> Self {
        self.data.rate_weight = Some(value);
        self
    }

    pub fn transform(mut self, transform: ParameterTransform) -> Self {
        self.data.parameter_transform = Some(transform);
        self
    }
}

impl TypeBuilder for FuncBuilder {
    fn build(self) -> Result<TypeId> {
        wasm::with_core(|c, s| c.call_funcb(s, &self.data))
    }
}

pub fn func<I, O>(inp: I, out: O, mat: u32) -> Result<FuncBuilder>
where
    I: TypeBuilder,
    O: TypeBuilder,
{
    Ok(FuncBuilder {
        data: TypeFunc {
            inp: inp.build()?,
            out: out.build()?,
            mat,
            ..Default::default()
        },
        ..Default::default()
    })
}

impl TypeBuilder for FuncParams {
    fn build(self) -> Result<TypeId> {
        func(self.inp, self.out, self.mat)?.build()
    }
}
